SYCL/AtomicRef/add_orders_scopes.cpp

```diff
@@ -0,0 +1,49 @@
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsycl-targets=%sycl_triple %s -o %t.out \
+// RUN: -Xsycl-target-backend --cuda-gpu-arch=sm_70
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+#define SYCL_USE_NATIVE_FP_ATOMICS
+
+#include "add.h"
+#include <iostream>
+using namespace sycl;
+
+template <typename T, typename Difference = T,
+          memory_order order = memory_order::relaxed>
+void add_test_scopes(queue q, size_t N) {
+  add_test<T, Difference, order, memory_scope::system>(q, N);
+  add_test<T, Difference, order, memory_scope::device>(q, N);
+  add_test<T, Difference, order, memory_scope::work_group>(q, N);
+  add_test<T, Difference, order, memory_scope::sub_group>(q, N);
+}
+
+template <typename T, typename Difference = T>
+void add_test_orders_scopes(queue q, size_t N) {
+  add_test_scopes<T, Difference, memory_order::relaxed>(q, N);
+  add_test_scopes<T, Difference, memory_order::acquire>(q, N);
+  add_test_scopes<T, Difference, memory_order::release>(q, N);
+  add_test_scopes<T, Difference, memory_order::acq_rel>(q, N);
+}
+
+int main() {
+  queue q;
+
+  constexpr int N = 32;
+  add_test_orders_scopes<int>(q, N);
+  add_test_orders_scopes<float>(q, N);
+  add_test_orders_scopes<unsigned int>(q, N);
+  add_test_orders_scopes<double>(q, N);
+  add_test_orders_scopes<long>(q, N);
+  add_test_orders_scopes<unsigned long>(q, N);
+
+  // Include long long tests if they are 64 bits wide
+  if constexpr (sizeof(long long) == 8) {
+    add_test_orders_scopes<long long>(q, N);
+    add_test_orders_scopes<unsigned long long>(q, N);
+  }
```

Comment:
The device associated with the queue must support `aspect::atomic64` for these tests to be valid. However, even if you check for that here the compiler may still try to generate the kernels for the 64-bit tests which it will try to build with the other kernels, even though it never intends to run them. This is the reason why many other tests have a `_atomic64` test variant.

Reply:
To handle all possible combinations of supported aspects each of these tests would have to be split into many files (for every combination of supported atomic64, 3 memory scopes and 4 memory orders). That would be a bit too many files in my opinion. 

In practice memory orders are currently only supported by sm_70 or higher NVidia devices. These also support all memory scopes and atomic64. That is why I decided to put all new tests into a single file that requires sm_70.

If you have a better suggestion how to handle this, I am happy to change it.

Reply:
I agree, the number of combinations becomes quite a headache, but the limitations of order/scope/atomic64 may apply to other devices than just CUDA ones.

An option would be to guard different combinations with macros you then define in different `//RUN` commands of the tests. This should make sure that the kernels are only generated for the combination test cases, avoiding compilation of invalid kernels by only running them if the device supports the combination of that test case. Does that make sense?

Reply:
That would still result in 24 RUN commands per file, but it does allow to merge all test files for each operation into just one.

Although I have to say I do not completely understand what mechanism skips the RUN lines for aspects that are unsupported in the hardware the test is running.

Reply:
That is a drawback of this approach; you cannot have e.g. XFAIL for each, but since you have to check for the features at runtime anyway the tests would have to dispatch.

The benefit of splitting them up (even through the macro alternative) is that the tests will not load binaries with unsupported instructions for a given device. That is, the individual RUNs would start out by checking that the corresponding order/scope/atomic64 is supported for the given device, and if not it should skip the test (with a print but not an error.)

Reply:
I still have to do runtime checks? That would result in these tests not being run until https://github.com/intel/llvm/pull/4853 is merged

Reply:
You have to do the runtime checks to make sure the device actually supports the features. CUDA is special in that regard as the SM version controls whether or not a feature is supported, but that's not the case for all devices.

> That would result in these tests not being run until intel/llvm#4853 is merged

Since the patch is up I don't see why we wouldn't wait for it. It would make the tests more robust from the get-go.

Reply:
I applied this kind of changes to tests for add. Can you check this is what you had in mind before I do the rest?

Reply:
That is exactly what I had in mind! It is quite a lot of RUN lines, but I suppose I got what was coming. Maybe it would make sense, just for good measure, to have a comment on each RUN grouping to have a short note specifying which test-configuration it is running.